Первое ДЗ. Приложение по чтению CSV файла из ресурсов

Greyzl opened the pull request.

Commit:
Первое ДЗ. Приложение по чтению CSV файла из ресурсов

stvort commented on hw01-intro/src/main/java/ru/otus/Main.java:
```diff
@@ -0,0 +1,20 @@
+    public static void main(String[] args) {
+        ApplicationContext ctx = new ClassPathXmlApplicationContext("spring/context.xml");
+        QuestionService questionService = ctx.getBean(QuestionService.class);
+        List<Question> questions = questionService.getQuestions();
+        for (Question question: questions){
+            System.out.println(question);
+        }
```
Бизнес-логику из main лучше вынести в отдельный сервис/раннер, оставив только получение бина и вызов тестирования. Зависимости бина лучше настроить в context.xml

stvort commented on hw01-intro/src/main/java/ru/otus/Main.java:
```diff
@@ -0,0 +1,20 @@
+        QuestionService questionService = ctx.getBean(QuestionService.class);
+        List<Question> questions = questionService.getQuestions();
+        for (Question question: questions){
+            System.out.println(question);
```
Работу с консолью (ввод/вывод, System.in/out) можно вынести в отдельный сервис. Так будет удобнее тестировать классы, где этот ввод/вывод используется. Причем System.in/out такому сервису можно будет тоже передать через конструктор. Такой сервис ничего не будет знать о предметной области и может быть переиспользован

stvort commented on hw01-intro/src/main/java/ru/otus/entity/enums/QuestionType.java:
```diff
@@ -0,0 +1,25 @@
+
+public enum QuestionType {
+
+    FREEFORMAT("freeformat");
```
Я на лекции просил не делать поддержку вопросов разных типов

stvort commented on hw01-intro/src/main/java/ru/otus/entity/Question.java:
```diff
@@ -0,0 +1,51 @@
+
+    private String prompt;
+
+    private String answer;
```
Потому что это лишнее усложнение. Что тут, например, будет содержаться при типе вопросов с несколькими правильным вариантами?

stvort commented on hw01-intro/src/main/java/ru/otus/dao/imp/QuestionDaoCSV.java:
```diff
@@ -0,0 +1,52 @@
+
+public class QuestionDaoCSV implements QueationDao {
+
+    private QuestionMapperCsv questionMapperCsv;
+
+    private String fileName ;
```
`final`

stvort commented on hw01-intro/src/main/java/ru/otus/dao/imp/QuestionDaoCSV.java:
```diff
@@ -0,0 +1,52 @@
+
+    }
+
+    public void setQuestionMapperCsv(QuestionMapperCsv questionMapperCsv) {
+        this.questionMapperCsv = questionMapperCsv;
+    }
+
+    public void setFileName(String fileName) {
+        this.fileName = fileName;
+    }
```
Мы обсуждали, что на курсе принято внедрение через конструктор и никак иначе

stvort commented on hw01-intro/src/main/java/ru/otus/dao/imp/QuestionDaoCSV.java:
```diff
@@ -0,0 +1,52 @@
+            questionsCsv.forEach(questionCsv ->
+                    questions.add(questionMapperCsv.mapQuestion(questionCsv)));
+        } catch (Exception e){
+            throw new RuntimeException(e);
```
Лучше свое сделать исключение, которое имеет смысл с т.з. предметной области

stvort commented on hw01-intro/gradlew.bat:
```diff
@@ -0,0 +1,91 @@
+@rem
```
Для сборки нужно, чтобы в репозитории проекта присутствовал файл `build.gradle`

Commit:
Поправил замечение
 1. добавил забытые файлы гредла
 2. Добавил сервис вывода
 3. Добавил исключения для обработки CSV
 4. Добавил final

Commit:
Удалил лишний пакет

stvort commented on hw01-intro/src/main/java/ru/otus/service/QuestionService.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface QuestionService {
+
+    List<Question> getQuestions();
+
+    void showQuestions();
```
Это нарушение SRP (**S**OLID). Тут явно больше одной причины для изменения класса

stvort commented on hw01-intro/src/main/java/ru/otus/service/imp/OutputServiceImp.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public <T> void output(List<T> objects) {
+        for (T object: objects){
+            printStream.println(object);
```
Сервис ввода-вывода лучше сделать именно отдельными и универсальным. Т.е. выводящим строки. Сейчас он тоже универсальный, но только за счет использования `toString`, об неприемлемости использования которого для вывода данных пользователю, мы говорили на занятии по введению в архитектуру

stvort commented on hw01-intro/src/main/java/ru/otus/service/imp/OutputServiceImp.java:
```diff
@@ -0,0 +1,23 @@
+        for (T object: objects){
+            printStream.println(object);
+        }
+        System.out.println();
```
`printStream`?

stvort commented on hw01-intro/src/main/java/ru/otus/mapper/QuestionMapperCsv.java:
```diff
@@ -0,0 +1,14 @@
+public class QuestionMapperCsv {
+
+    public Question mapQuestion(QuestionCsv questionCsv) {
+        Question question = new Question();
+        question.setText(questionCsv.getText().trim());
+        question.setPrompt(questionCsv.getQuestionVariants().get(0).trim());
```
1. Конструктор лучше сеттеров. И лаконичнее
2. А что если нет ни одного варианта? Мы поймаем это исключение?

stvort commented on hw01-intro/src/main/resources/testQuestions.csv:
```diff
@@ -0,0 +1,6 @@
+text,parameter1
+What is your name?, enter your name
```
У нас система тестирования, а не опрос. Плюс сообщения системы не часть теста. Так что если "enter your name" это не правильный ответ на вопрос "What is your name?", то нужно что-то поменять в тесте, чтобы было по итогу можно в следующем ДЗ оценить результаты тестирования

Commit:
Много изменений:
1. Попробовал заняться SRP и выделил сервис вывода отдельно, добавил Formatter. Добавил запускатор и процессор для вывода вопросов
2. Постарался перейти от сеттеров к коструктору
3. Поменял значения в файле с вопросов на тест

stvort commented on hw01-intro/src/main/java/ru/otus/processor/Processor.java:
```diff
@@ -0,0 +1,5 @@
+package ru.otus.processor;
+
+public interface Processor {
+    void process();
+}
```
Что именно процессит этот процессор?

Greyzl replied:
Поправил название, но не уверен что оно стало лучше отражать суть. Данный процессор выдает разные сценарии использования теста. Вывод вопросов без сбора ответов один из сценариев использования

stvort commented on hw01-intro/src/main/java/ru/otus/dao/imp/QuestionDaoCSV.java:
```diff
@@ -0,0 +1,46 @@
+    public List<Question> loadQuestions() {
+        List<Question> questions = new ArrayList<>();
+        try {
+            InputStreamReader streamReader = new InputStreamReader(
+                    new ClassPathResource(fileName).getInputStream());
```
Лучше воспользоваться try-with-resources для закрытия ресурса после использования

stvort commented on hw01-intro/src/main/java/ru/otus/dao/imp/QuestionDaoCSV.java:
```diff
@@ -0,0 +1,46 @@
+            questionsCsv.forEach(questionCsv ->
+                    questions.add(questionMapperCsv.mapQuestion(questionCsv)));
+        } catch (IllegalStateException | IOException  e){
+            throw new CvsFileReadException(e);
```
Лучше конечно безотносительно источника данных исключение сделать. Так его можно будет обработать выше по типу. Сейчас нельзя без нарушения SOLID

stvort commented on hw01-intro/src/main/java/ru/otus/entity/Question.java:
```diff
@@ -0,0 +1,22 @@
+public class Question {
+    private final String text;
+
+    private final List<String> answerOptions;
```
Ответ же может быть правильным или не правильным?

stvort commented on hw01-intro/src/main/resources/testQuestions.csv:
```diff
@@ -0,0 +1,7 @@
+text,answer_option_1,answer_option_2,answer_option_3
+What is your name?,,,
```
Еще раз. Это сообщение часть приложения, а не теста

Greyzl committed:
1. Перенес обращение к стриму в try-with-resourses
2. Перенаимновал Exception из CSV в LoadQuestionsException
3. Создал сервис по составлению вопросов о пользователе
4. Перенаимновал интерфейс процессора в TestProcessor

Commit:
Добавил обработку "правильного ответа" из файла

stvort commented on hw01-intro/src/main/java/ru/otus/processor/impl/PrintTestQuestionsProcessor.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public void process() {
+        Question userNameQuestion = userQuestionService.getUserNameQuestion();
+        String formattedUserNameQuestion = outputFormatter.format(userNameQuestion);
+        outputService.output(formattedUserNameQuestion);
```
Еще раз) Есть элементы интерфейса, сообщения системы и т.д. Они **не обязаны быть вопросами теста ни в каком их проявлении**. Иначе, что бы мы делали если бы у нас было приложение не по тестированию, а ФИО нужно было бы все равно спросить? Просто `UserInfo userInfo = userService.getUserInfo();`, где внутри метода выводим тупо строкой запрос имени и фамилии, читаем ответ, делаем из него объект и возвращаем наружу

stvort commented on hw01-intro/src/main/java/ru/otus/dao/impl/QuestionDaoCSV.java:
```diff
@@ -0,0 +1,45 @@
+                    .parse();
+            questionsCsv.forEach(questionCsv ->
+                    questions.add(questionMapperCsv.mapQuestion(questionCsv)));
+        } catch (IllegalStateException | IOException  e){
```
А еще желательно поймать `QuestionMapException`

Commit:
1. Добавил сервис пользователей и вывод вопроса через него
2. Добавил прокидываение QuestionMapException через LoadQuestionsException

stvort commented on hw01-intro/src/main/java/ru/otus/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,20 @@
+        this.outputService = outputService;
+    }
+    
+    @Override
+    public void getUser() {
+        outputService.output(USERNAME_QUESTION);
+    }
```
И все же лучше, чтобы метод, который начинается на `get`, что-то возвращал. Поскольку следующее слово это 'User', то его он и должен вернуть.

>Не хотел вводить новую сущность до того как начну делать сбор ответов. Казалось это не имело смысла

Запрос имени пользователя, это часть следующей работы. В первой этого делать было не надо. Но раз сделали, то сделать надо полностью и максимально хорошо

stvort commented on hw01-intro/src/main/java/ru/otus/entity/Question.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final String rightAnswer;
+
+    private final List<String> answerOptions;
```
>Ответ же может быть правильным или не правильным?

Не исправлено. Тут имеется в виду, что согласно ДЗ мы
>отдаём предпочтение объектам и классам, а не строчкам и массивам/спискам строчек

А значит это должен быть список объектов "Ответ" с признаком правильности и текстом внутри
